Add Uniswap V3 Arbitrum trades

mtitus6 opened the pull request:
Brief comments on the purpose of your changes:


*For Dune Engine V2*
I've checked that:
General checks:
* [X] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [X] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [X] the filename is unique and ends with .sql
* [X] each sql file is a select statement and has only one view, table or function defined  
* [X] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)

Join logic:
* [X] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [X] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [X] where block_time >= date_trunc("day", now() - interval '1 week')
* [X] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [X] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

jeff-dude commented:
please add ```arbitrum``` directory into the YAML structure of dbt_project here:
https://github.com/duneanalytics/spellbook/blob/main/dbt_project.yml#L85

this addition allows you to assign default schema name & materialization for any models within the ```models/uniswap/arbitrum/``` directory

jeff-dude commented on models/uniswap/arbitrum/uniswap_arbitrum_schema.yml:
```diff
@@ -0,0 +1,132 @@
+      - &token_bought_address
+        name: token_bought_address
+        description: "Contract address of the token bought"
+        tests:
```
in order for this test to run, you need to manually add rows to our dex trades seed file:
https://github.com/duneanalytics/spellbook/blob/main/seeds/dex/trades/dex_trades_seed.csv

uniswap has a linked analytics page on their app (that's how i did for ethereum version), you can find example tx's there where you can open on etherscan and fill out rows as needed

the goal is to ensure the columns pulled for bought/sold are applied correctly & not flipped (common issue on v1 engine)

mtitus6 replied:
I added 3 transactions to the seed file. USDC to WETH , WETH to USDC, and USDC to DAI

jeff-dude commented on models/uniswap/arbitrum/uniswap_arbitrum_trades.sql:
```diff
@@ -0,0 +1,33 @@
+{{ config(
+        alias ='arbitrum_trades'
```
the ```alias``` can  simply be ```trades``` as the schema is defined in the dbt_project file, which will show ```arbitrum``` in the name (once you apply the changes i mentioned in other comment)

mtitus6 replied:
Gotcha.  It wouldn't compile which is why I changed it initially, but it compiled OK after changing the dbt_project file.

jeff-dude commented on models/uniswap/arbitrum/uniswap_arbitrum_trades.sql:
```diff
@@ -0,0 +1,33 @@
+
+SELECT *
+FROM
+(
```
may be missing closing ```)``` at the end?

mtitus6 replied:
Got it. Thanks

mtitus6 commented:
All of the feedback items have been resolved in the latest commits.

jeff-dude commented on models/uniswap/arbitrum/uniswap_v3_arbitrum_trades.sql:
```diff
@@ -0,0 +1,110 @@
+    {% endif %}
+LEFT JOIN {{ ref('tokens_erc20') }} erc20a
+    ON erc20a.contract_address = dexs.token_bought_address 
+    AND erc20a.blockchain = 'arbitrum'
```
there doesn't appear to be an erc20 tokens view for arbitrum:
```
select blockchain, count(1)
from tokens.erc20
group by blockchain
order by 1 desc
```

here is code for other blockchains:
https://github.com/duneanalytics/spellbook/tree/main/models/tokens

with this info missing, the join returns nothing, then all fields in the select above using either ```erc20a``` or ```erc20b``` return nulls.

example:
```
select
  *
from
  dbt_jeff_uniswap_v3_arbitrum.trades
where
  block_date >= '2022-10-10'
```

are you comfortable trying to add that model for arbitrum erc20 tokens?

mtitus6 replied:
I would certainly be up to the challenge but definitely need a nudge in the right direction.  They are all hard coded so there must be a method to grabbing them.  Looking into Arbiscan's API now but not seeing anything obvious.

mtitus6 replied:
Went ahead and started just pulling them manually.  Did not take as long as I expected.  I did the first 50 by market cap.

jeff-dude replied:
thank you for the addition! i wanted to share one last thing that may help prior to moving towards the merge.

here is a query (stolen from another useful wizard) for tracking missing tokens in the view(s):
*note since arbitrum is starting from scratch, the missing token portion is commented out
https://dune.com/queries/1399945

if you'd like to take a look and consider adding any more of the most used tokens on uniswap v3 for arbitrum (this is just in the last week, you could open up further date wise), feel free to do so

jeff-dude replied:
in a similar fashion, for most complete ```amount_usd``` values, this file is maintained in a similar fashion:
https://github.com/duneanalytics/spellbook/blob/main/models/prices/prices_tokens.sql

you can add the same subset of tokens to this file, it's a slightly different format but same idea

note that the ```token_id``` column is important for accuracy there, as it tells us how to call the coinpaprika API. that is where you can validate the id. here is helpful link:
https://api.coinpaprika.com/v1/coins/

addresses can be searched and id's found

you are the first to add arbitrum dex trades data, so it comes with a bit more work sadly 😅

mtitus6 replied:
Fun! I had no idea I was headed down such a rabbit hole.  That's exciting though that I can help lay the foundation to keep building Arbi out.  Thanks for the query.  I reworked it a bit so it will only return addresses I had not already added.  Maybe someone will find it useful later -->[https://dune.com/queries/1401328].  
Will work on prices tomorrow.  Thanks for sharing the API.

jeff-dude replied:
sure thing, look forward to seeing the add!

jeff-dude commented on seeds/dex/trades/dex_trades_seed.csv:
```diff
@@ -19,4 +19,7 @@ ethereum,Curve,Regular,2022-09-03,0x0fc5910c411bdbbf1827fa278c22b3351892e8931e7c
-ethereum,Curve,Factory V2,2022-09-03,0x02fe546401dd52e07528532ed1ac4ed1d4f525118836f71a4014282fb6a06839,4,0xaf5191b0de278c7286d6c7cc6ab6bb8a73ba2cd6,51356173218735436675801,0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48,33333000000
\ No newline at end of file
+ethereum,Curve,Factory V2,2022-09-03,0x02fe546401dd52e07528532ed1ac4ed1d4f525118836f71a4014282fb6a06839,4,0xaf5191b0de278c7286d6c7cc6ab6bb8a73ba2cd6,51356173218735436675801,0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48,33333000000
+arbitrum,uniswap,3,2022-10-11,0x88a863ee12e4258c93bf41651853093f071d53facf0cbe0ae7b2bad040928f50,3,0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8,3198364498,0x82aF49447D8a07e3bd95BD0d56f35241523fBab1,2500000000000000000
```
looks like you may have flipped token bought/sold here? or is it flipped in the code/model?
```
with
  unit_test as (
    select
      lower(m.token_bought_address) as model_token_bought,
      lower(seed.token_bought_address) as seed_token_bought,
      lower(m.token_sold_address) as model_token_sold,
      lower(seed.token_sold_address) as seed_token_sold,
      case
        when lower(m.token_bought_address) = lower(seed.token_bought_address) then True
        else False
      end as token_bought_test
    from
      dbt_jeff_uniswap_v3_arbitrum.trades m
      inner join dbt_jeff_test_data.dex_trades_seed seed on m.tx_hash = seed.tx_hash
      and m.evt_index = seed.evt_index
      and m.block_date = seed.block_date
      and m.blockchain = seed.blockchain
      and m.project = seed.project
      and m.version = seed.version
  )
select
  *
from
  unit_test
where
  token_bought_test = False
```

this is why we try to add the tests 👍

jeff-dude replied:
also addresses should be lower everywhere (they aren't in the seed here). to be safe, i've added lower() function in the generic test applied

mtitus6 replied:
I had them flipped in the seed file.  I swapped them and made lower case

mtitus6 commented:
Modified the dex_trades_seed.csv file switching the bought and sold columns, and making contract addresses lower case.  Added the Arbitrum erc20.tokens table with the first 50 tokens by market cap.

jeff-dude commented:
other than my two comments above, everything is looking great! look forward to more tokens added and getting arbitrum data going into dex trades

dune-eng commented:
Workflow run id 3246759461 approved.

dune-eng commented:
Workflow run id 3257691383 approved.

mtitus6 commented:
Updated the prices model with all applicable arbitrum tokens.

jeff-dude commented:
i think i'm comfortable with where this stands -- great work! thank you for the patience and the back and forth. if you use this spell in dashboards & notice any issues, always feel free to open new PRs with enhancements. i think adding tokens / prices will be common, then we can always do full refresh for historical trades to reflect the adds

mtitus6 commented:
Awesome.  No worries and thanks for your helpful guidance ☺️.  This was my first PR ever 🎉 and I look forward to contributing more.

mtitus6 commented:
> Well done @mtitus6 👏

Thanks @hagaetc.  Appreciate you taking time out of your day to say that ☺️.
